result.nim

```diff
@@ -0,0 +1,352 @@
+
+    case isOk*: bool
+    of false:
+      error*: E
+    of true:
+      value*: T
+
+  # TODO cannot raise generic error, so cannot include error value :(
+  # https://github.com/nim-lang/Nim/issues/7845
+  ResultError = object of Exception
+
+# TODO check that r is a Result
+proc ok*(r: typedesc, v: auto): r =
+  ## Initialize a result with a success and value
+  ## Example: `Result[int].ok(42)`
+  r(isOk: true, value: v)
+proc ok*[T, E](self: var Result[T, E], v: auto) =
+  ## Set the result to success and update value
+  ## Example: `result.ok(42)`
+  self = Result[T, E].ok(v)
+
+proc err*(r: typedesc, e: auto): r =
+  ## Initialize the result to an error
+  ## Example: `Result[int].err("uh-oh")`
+  r(isOk: false, error: e)
+
+proc err*[T, E](self: var Result[T, E], v: auto) =
+  ## Set the result as an error
+  ## Example: `result.err("uh-oh")`
+  self = Result[T, E].err(v)
+
+template isErr*[T, E](self: Result[T, E]): bool = not self.isOk
+
+proc map*[T, E, A](self: Result[T, E], f: proc(x: T): A {.closure.}): Result[A, E] =
+  ## Transform value using f, or return error
+  if self.isOk: result.ok(f(self.value))
+  else: result.err(self.error)
+
+proc mapErr*[T, E, A](self: Result[T, E], f: proc(x: E): A {.closure.}): Result[T, A] =
+  ## Transform error using f, or return value
+  if self.isOk: result.ok(self.value)
+  else: result.err(f(self.error))
+
+proc mapConvert*[T0, E0](self: Result[T0, E0], T1: typedesc): Result[T1, E0] =
+  ## Convert result value to A using an implicit conversion
+  ## Would be nice if it was automatic...
+  if self.isOk: result.ok(self.value)
+  else: result.err(self.error)
+
+proc mapCast*[T0, E0](self: Result[T0, E0], T1: typedesc): Result[T1, E0] =
+  ## Convert result value to A using a cast
+  ## Would be nice with nicer syntax...
+  if self.isOk: result.ok(cast[T1](self.value))
+  else: result.err(self.error)
+
+template `and`*[T, E](self: Result[T, E], other: untyped): untyped =
+  ## Evaluate `other` iff self.isOk, else return error
+  ## fail-fast - will not evaluate other if a is an error
+  if self.isOk:
+    other
+  else:
+    type R = type(other)
+    R.err(self.error)
+
+template `or`*[T, E](self: Result[T, E], other: untyped): untyped =
+  ## Evaluate `other` iff not self.isOk, else return self
+  ## fail-fast - will not evaluate other if a is a value
+  if self.isOk: self
+  else: other
+
+template `+`*[T, E](self, other: Result[T, E]): untyped =
+  ## Perform `+` on the values of self and other, if both are
+  type R = type(other)
+  if self.isOk:
+    if other.isOk:
+      R.ok(self.value + other.value)
+    else:
+      R.err(other.error)
+  else:
+    R.err(self.error)
+
+template catch*(body: typed): Result[type(body), ref Exception] =
+  ## Convert a try expression into a Result
+  type R = Result[type(body), ref Exception]
+
+  try:
+    R.ok(body)
+  except:
+    R.err(getCurrentException())
+
+template capture*(a: typedesc, e: ref Exception): Result[a, ref Exception] =
```

Comment:
Perhaps this API should be optimized to take advantage of the `result` symbol in the proc.

```nim
proc foo: Result[int, ref Exception] =
   if errorDetected:
      return captureStack() # this can figure out all the details from the `result` var.
```

Reply:
hm, as an overload perhaps - there's the case where you want to add some custom information to the error, or where you return the base class of a hierarchy of errors, but raise something more specific

Reply:
Yes, these short-cut forms could be just overloads.

Reply:
added a test for it